### ISSUE TITLE
SQLAlchemy can't handle reflections for tables with no columns - (raising "NoSuchTableError")

### ISSUE THREAD

**Issue by bohemia420** [sqlalchemy/sqlalchemy#7117]
### Describe the bug

The bug being SQLAlchemy being unable to handle/provide reflection of tables with no columns, i.e call to :

```Table(name, metadata, schema=schema, autoload_with=autoload_with, extend_existing=True)```
would result in a NoSuchTableException being raised, should the table exist but with no columns. 

SQLAlchemy's `engine/reflection.py` i.e `lib/sqlalchemy/engine/reflection.py` seems to recognise i.e "find" tables with the `table_found` variable, **only through iterating over its columns**. Since an empty table wouldn't have any, the `table_found` stays False and results in "NoSuchTableError" Exception being thrown:

file : `sqlalchemy/engine/reflection.py`
```
        found_table = False
        cols_by_orig_name = {}

        for col_d in self.get_columns(
            table_name, schema, **table.dialect_kwargs
        ):
            found_table = True

            self._reflect_column(
                table,
                col_d,
                include_columns,
                exclude_columns,
                cols_by_orig_name,
            )

        if not found_table:
            raise exc.NoSuchTableError(table.name)
```

### To Reproduce

```python
upgrading the version to `1.4.23`, even `1.4.25` doesn't work

# pip3.9 freeze | grep -i sqlalchemy
SQLAlchemy==1.4.23
```

and execute the following steps in python3 shell:

```
# python3.9
Python 3.9.7 (default, Sep  3 2021, 20:19:49) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sqlalchemy import create_engine, Table, Column, Integer, String, MetaData
>>> engine = create_engine('postgresql://<user>:<pass>@<db>:5432/<database>')
>>> engine.table_names()
<stdin>:1: SADeprecationWarning: The Engine.table_names() method is deprecated and will be removed in a future release.  Please refer to Inspector.get_table_names(). (deprecated since: 1.4)
[...]
>>> metadata = MetaData(bind=engine)
>>> emp23 = Table('emp23', metadata, )
>>> tab23 = Table('tab23', metadata,Column('id', Integer, primary_key = True), )
>>> metadata.create_all(engine)
>>> engine.table_names()
['emp23', 'tab23', ...]
>>> Table('tab23', MetaData(bind=engine), schema='public', autoload_with=engine, extend_existing=True)
Table('tab23', MetaData(bind=Engine(postgresql://<user>:***@<db>:5432/<database>)), Column('id', INTEGER(), table=<tab23>, primary_key=True, nullable=False, server_default=DefaultClause(<sqlalchemy.sql.elements.TextClause object at 0x7efeaf9ba580>, for_update=False)), schema='public')
>>> Table('emp23', MetaData(bind=engine), schema='public', autoload_with=engine, extend_existing=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 2, in __new__
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/deprecations.py", line 298, in warned
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 601, in __new__
    metadata._remove_table(name, schema)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 596, in __new__
    table._init(name, metadata, *args, **kw)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 671, in _init
    self._autoload(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 706, in _autoload
    conn_insp.reflect_table(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/reflection.py", line 788, in reflect_table
    raise exc.NoSuchTableError(table.name)
sqlalchemy.exc.NoSuchTableError: emp23
>>> 
```

### Error

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 2, in __new__
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/deprecations.py", line 298, in warned
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 601, in __new__
    metadata._remove_table(name, schema)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 596, in __new__
    table._init(name, metadata, *args, **kw)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 671, in _init
    self._autoload(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/schema.py", line 706, in _autoload
    conn_insp.reflect_table(
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/reflection.py", line 788, in reflect_table
    raise exc.NoSuchTableError(table.name)
sqlalchemy.exc.NoSuchTableError: emp23
``` 

### Versions

- OS:Docker Image atop `FROM python:3-buster` ; Darwin Kernel Version 20.2.0
- Python: Python 3.9.7
- SQLAlchemy: 1.4.18; 1.4.23; 1.4.25;
- Database: postgres (Dockerized)
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): psycopg2

### Additional context

This issue has to be agnostic of the linux/python/dbclient version etc. and is mostly due to how `found_table` is being inferred while providing table reflections. I have a solution I'm PRing next, and it involves referring `engine.table_names`, as follows:

```
 771         found_table = True if table.name in self.engine.table_names() else False
 772         cols_by_orig_name = {}
 773 
 774         for col_d in self.get_columns(
 775             table_name, schema, **table.dialect_kwargs
 776         ):
 777             self._reflect_column(
 778                 table,
 779                 col_d,
 780                 include_columns,
 781                 exclude_columns,
 782                 cols_by_orig_name,
 783             )
 784 
 785         if not found_table:
 786             raise exc.NoSuchTableError(table.name)

```

TL;DR - making use of `self.engine.table_names()` and checking if `table.name` is in there. This ways, even empty tables will get through, and not remain FNs anymore.

**Comment by CaselIT** [comment]
Hi,

I agree that the current behaviour is unfortunate, but what's the use case for a table with no columns?

Also `engine.table_names()` is deprecated, as the warning indicates, so it can't be used for this

**Comment by bohemia420** [comment]
> I agree that the current behaviour is unfortunate, but what's the use case for a table with no columns?

great question, @CaselIT . Actually there's an opensource DB-interacting project `https://github.com/centerofci/mathesar` I'm beginning to contribute too. It allows one to create tables, wherein columns can also be deleted. 

Problem thus arises (and what defines this not so trivial usecase) when even the last column is deleted and loading the home page to retrieve (reflections of) tables fails with "NoSuchTableError" for the table that now had its last column deleted, above.

I know it may be absurd to have/work with tables that have no columns, at all, thus two solutions look promising:

- SQLAlchemy honoring the reflection of tables with no columns, probably as per the solution recommended in this PR (reiterating).
- We explicitly handling the case where one tries to drop the last column in the table - accordingly either refraining the last column replaced or deleting the table along, recursively (As the last column goes off!)

**Comment by CaselIT** [comment]
I think the first option is valid, meaning that the table exists but sqlalchemy is reporting no-such-table error, so that's a bug.

I think it's fine fixing the issue in sqlalchemy

**Comment by zzzeek** [comment]
please track this on the original issue which is #3247.
